src/ordering/service.py

```diff
@@ -1,8 +1,53 @@
+import logging
 from uuid import UUID
 
+from injector import inject
+
+from application.settings import Settings
+from currency import BTC, ExchangeRateService
+
 from .commands import CreateBuyOrder
+from .db import Repository
+from .errors import BalanceLimitExceeded, OrderAlreadyExists
+from .events import BuyOrderCreated
 
+log = logging.getLogger(__name__)
 
+
+@inject
 class Service:
+    def __init__(
+            self,
+            settings: Settings,
+            repository: Repository,
+            exchange_rates: ExchangeRateService,
+    ) -> None:
+        self._bought_btc_limit = settings.ordered_btc_limit
+        self._repository = repository
+        self._get_order_id = repository.get_order_id
+        self._get_btc_rate = exchange_rates.get_bitcoin_rate
+
     def create_buy_order(self, command: CreateBuyOrder) -> UUID:
-        raise NotImplementedError
+        log.info(command)
+        order_id = self._get_order_id(command.id)
+        if order_id is not None:
+            raise OrderAlreadyExists(order_id)
+
+        rate = self._get_btc_rate(command.currency)
+        btc = BTC(command.amount / rate.price)
+
+        with self._repository.lock() as current_balance:
+            if current_balance + btc > self._bought_btc_limit:
+                raise BalanceLimitExceeded(self._bought_btc_limit)
+
+            order = self._repository.create(
```

Comment:
I don't like how repository is burdened with this responsibility. The way I understand repositories, they are meant only to provide persistence-oriented stuff like save and get OR colleciton-like interface with get, add but no save. 

Your implementation does neither. So it does not resemble a repository pattern and that surprises me.
I'd probably create a factory for that. 

Also, I think you're trying to hard to kill two birds with one stone. IMO there's either one aggregate with all orders and balance OR two separate aggregates - one is wallet balance and the second one is the order. Hence, I'm bit surprised by seeing all these locking of balance using orders repository 🤔

Reply:
Actually, this is another topic that I discussed before and heard that it's not so clear. This is not a shortcut that I've made here, it was a choice that I've done.
But this is also an interesting topic to discuss.